src/logic/small.lean

```diff
@@ -94,4 +94,25 @@ end
 instance small_of_encodable (α : Type v) [encodable α] : small.{w} α :=
 small_of_injective _ (encodable.encode_injective)
 
+instance small_Pi {α} (β : α → Type*) [small.{w} α] [∀ a, small.{w} (β a)] :
+  small.{w} (Π a, β a) :=
+⟨⟨Π a' : shrink α, shrink (β ((equiv_shrink α).symm a')),
+  ⟨equiv.Pi_congr (equiv_shrink α) (λ a, by simpa using equiv_shrink (β a))⟩⟩⟩
+
+instance small_sigma {α} (β : α → Type*) [small.{w} α] [∀ a, small.{w} (β a)] :
+  small.{w} (Σ a, β a) :=
+⟨⟨Σ a' : shrink α, shrink (β ((equiv_shrink α).symm a')),
+  ⟨equiv.sigma_congr (equiv_shrink α) (λ a, by simpa using equiv_shrink (β a))⟩⟩⟩
+
+instance small_prod {α β} [small α] [small β] : small (α × β) :=
```

Comment:
Maybe it's best to always explicitly write the universe, i.e. `small.{w}`, even when Lean fills things in correctly by itself?

Reply:
This currently produces the conclusion `small_prod.{u_1 u_2 u_3 u_4} : small.{(max u_3 u_4) (max u_1 u_2)}`. Shouldn't the conclusion be `small_prod.{u, v, w} : small.{w}`? I feel like the instances themselves should always be `small.{v}`, since `small_max` can always synthesize `small.{max u v}` from such a term.